### ISSUE TITLE
Generic with bound unassignable to its bound after widening and narrowing

### ISSUE THREAD

**Issue by AnJunHao** [astral-sh/ty#3137]
### Summary

Example code:

```python

class A: ...

def widen[T](param: T) -> T | A:
    return A()

def widen_include_none[T](param: T) -> T | None:
    return None

class SomeClass[T: int | str]:
    field: T

    def case_ok(self):
        test: int | str
        test = self.field

    def case_a(self):
        test: int | str
        assert not isinstance(self.field, int)
        test = self.field # reports Object of type `T@SomeClass & ~int` is not assignable to `int | str` (invalid-assignment) [Ln 19, Col 16]


    def case_b(self):
        test: int | str
        widened_field = widen(self.field)
        assert not isinstance(widened_field, A)
        test = widened_field # reports Object of type `T@SomeClass & ~A` is not assignable to `int | str` (invalid-assignment) [Ln 25, Col 16]

    def case_c(self):
        test: int | str
        widened_field = widen_include_none(self.field)
        assert widened_field is not None
        test = widened_field # fine

```

After narrowing, the checker infers types such as T & ~int and T & ~A. Since T is bounded by int | str, these narrowed types should still be subtypes of int | str and therefore assignable to that bound. The current behavior appears to lose the original bound relationship after widening/narrowing, causing valid assignments to be rejected.

See playground for repro:
https://play.ty.dev/32810079-ca47-4922-bf64-f6a1a42af8f6

### Version

ty 0.0.25 (also seen on previous versions)

**Comment by carljm** [comment]
Hm, that's odd. Thanks for the report!

**Comment by sharkdp** [comment]
A simplified case would be
```py


def _[T: int | str](x: T):
    a: int | str = x  # okay

    assert not isinstance(x, str)

    b: int | str = x  # error
```

I think this is related to https://github.com/astral-sh/ty/issues/2726

**Comment by carljm** [comment]
@sharkdp Yes, I think it's the same root issue, although (unless we can come up with a more general approach) I think the fixes will likely be separate. The core problem is that we need to flatten-and-simplify intersections (and maybe unions) containing typevars in order for various operations to properly account for the intersection, but we can't eagerly do this simplification or we lose the association with the typevar type itself. We already do this flatten-and-simplify in comparisons, this issue needs it in relations, #2726 needs it in attribute access.

**Comment by carljm** [comment]
The mypy approach I described in https://github.com/astral-sh/ty/issues/2726#issuecomment-3854997211 (synthesize a new typevar with narrowed bound, but retain its association with the original typevar) is a more general approach that could potentially work for us.
